### DISCUSSION TITLE
File Picker: Uri vs String

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#11003] [General]
When it comes to Urls, if you use them as strings, Roslyn analyzers warn to use Uri instead of String.

When it comes to file paths, does the same apply? I tried with that assumption, to realize that `System.IO.File.OpenRead(filePath)` does not take Uri, and does not recognize the `file:///` syntax.

This guidelines applies to Urls only, not file paths. So where did I get that idea to store file paths as Uri?

Oh right, Avalonia v11 file picker handling. Is using Uri here the right design decision?

When I get to use Uri to actually open a file, I need to convert it back to a string.

`filePath.ToString()` returns `file:///mypath/myfile.txt` and is not recognized by `OpenRead`

`filePath.LocalPath` returns the standard `C:\mypath\myfile.txt` but may not be supported by all devices.

So what's the right way to do it?

**Comment by maxkatz6** [comment]
> When it comes to file paths, does the same apply? I tried with that assumption, to realize that System.IO.File.OpenRead(filePath) does not take Uri, and does not recognize the file:/// syntax.

Uri.LocalPath can be used.
Or [IStorageItem.TryGetLocalPath](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Platform/Storage/StorageProviderExtensions.cs#L45-L55) as well which is internally the same, but for the whole path.

> Oh right, Avalonia v11 file picker handling. Is using Uri here the right design decision?

Yes, since there is more than just "file:". It can be "content:". Or it can be a relative file path.

> When I get to use Uri to actually open a file, I need to convert it back to a string.

You need to use OpenWrite or OpenRead if you want to keep it crossplatform. On desktop, though, you can convert it back to string, sure.

**Reply by mysteryx93** [reply]
With other platforms, are any direct file access via `System.IO` supported or not at all?

For `file:` or `content:`, I can't expect to open those with standard IO functions.

The case I have right now, storing settings file in the app storage folder. That one remains a standard absolute or relative path that should not cause problem?

At least; one benefit of storing picked files as Uri is that you'll know not to use them for standard IO functions.

**Reply by maxkatz6** [reply]
> With other platforms, are any direct file access via System.IO supported or not at all?

Win, Lin: yes.

macOS: yes, but OS might deny access if app sandboxed (same in Avalonia right now, https://github.com/AvaloniaUI/Avalonia/pull/6540 PR needs to be redone with new API)

Browser: no at all

Android: yes, same as macOS, except we have bookmarks API, so it's easier

iOS: yes, same as Android. Except, it also requires ["safe scope"](https://github.com/AvaloniaUI/Avalonia/blob/master/src/iOS/Avalonia.iOS/Storage/IOSSecurityScopedStream.cs#L22) for reading files. It is handled automatically with OpenRead/OpenWrite.

> For file: or content:, I can't expect to open those with standard IO functions.

You can try, if it's "file:". If user and OS gives you permissions, it will succeed. 

> The case I have right now, storing settings file in the app storage folder. That one remains a standard absolute or relative path that should not cause problem?

How is it related to the file picker? Or does user pick this app storage folder?
